pkg/controller/vitessshard/reconcile_rollout.go

```diff
@@ -0,0 +1,137 @@
+	"vitess.io/vitess/go/vt/topo/topoproto"
+
+	planetscalev2 "planetscale.dev/vitess-operator/pkg/apis/planetscale/v2"
+	"planetscale.dev/vitess-operator/pkg/operator/results"
+	"planetscale.dev/vitess-operator/pkg/operator/rollout"
+	"planetscale.dev/vitess-operator/pkg/operator/vttablet"
+)
+
+func (r *ReconcileVitessShard) reconcileRollout(ctx context.Context, vts *planetscalev2.VitessShard) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
+
+	if !rollout.Cascading(vts) {
+		// If the shard is not scheduled for a cascading update, silently bail out and do nothing.
+		return resultBuilder.Result()
+	}
+
+	podList := &v1.PodList{}
+	listOpts := &client.ListOptions{
+		Namespace: vts.Namespace,
+		LabelSelector: apilabels.Set(map[string]string{
+			planetscalev2.ComponentLabel: planetscalev2.VttabletComponentName,
+			planetscalev2.ClusterLabel:   vts.Labels[planetscalev2.ClusterLabel],
+			planetscalev2.KeyspaceLabel:  vts.Labels[planetscalev2.KeyspaceLabel],
+			planetscalev2.ShardLabel:     vts.Spec.KeyRange.SafeName(),
+		}).AsSelector(),
+	}
+
+	if err := r.client.List(ctx, listOpts, podList); err != nil {
+		return resultBuilder.Error(err)
+	}
+
+	// Safety checks and rolling updates must be deterministically ordered, so we access and sort pods by tablet alias.
+	tabletPods := make(map[string]*v1.Pod)
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		tabletAlias := vttablet.AliasFromPod(pod)
+		tabletKey := topoproto.TabletAliasString(&tabletAlias)
+		tabletPods[tabletKey] = pod
+	}
+
+	tabletKeys := make([]string, 0, len(vts.Status.Tablets))
+	for key := range vts.Status.Tablets {
+		tabletKeys = append(tabletKeys, key)
+	}
+	sort.Strings(tabletKeys)
+
+	for _, tabletKey := range tabletKeys {
+		tablet := vts.Status.Tablets[tabletKey]
+		if tablet.Available != v1.ConditionTrue {
+			// If any tablets are unhealthy, we should bail and not perform a rolling restart.
+			r.recorder.Eventf(vts, v1.EventTypeNormal, "RolloutPaused", "Waiting for tablet %v to be Available.", tabletKey)
+			return resultBuilder.Result()
+		}
+
+		pod, ok := tabletPods[tabletKey]
+		if !ok {
+			r.recorder.Eventf(vts, v1.EventTypeNormal, "RolloutPaused", "Waiting for desired tablet %v to be created.", tabletKey)
+			return resultBuilder.Result()
+		}
+
+		if rollout.Released(pod) {
+			// If any tablet has already been released, we should wait until it is finished to release another one.
+			r.recorder.Eventf(vts, v1.EventTypeNormal, "RolloutPaused", "Waiting for tablet %v to finish release.", tabletKey)
+			return resultBuilder.Result()
+		}
+	}
+
+	// Retrieve tablet pod to be released during this reconcile.
+	tabletKey, pod := getNextScheduledTablet(tabletKeys, tabletPods)
+	if tabletKey == "" {
+		// If we have no more scheduled tablets, uncascade the shard.
+		if err := r.uncascadeShard(ctx, vts); err != nil {
+			r.recorder.Eventf(vts, v1.EventTypeWarning, "UncascadeFailed", "Client failed to mark shard as uncascading.")
```

Comment:
Let's make this easier for users to understand, and also log the error:

```suggestion
			r.recorder.Eventf(vts, v1.EventTypeWarning, "UncascadeFailed", "Failed to mark cascading shard rollout as complete: %v", err)
```